src/main/java/net/minestom/server/network/packet/server/play/ParticlePacket.java

```diff
@@ -4,24 +4,70 @@
+    public ParticlePacket(int particleId, boolean longDistance,
+                          double x, double y, double z,
+                          float offsetX, float offsetY, float offsetZ,
+                          float maxSpeed, int particleCount, ParticleData data) {
+        this.particleId = particleId;
+        this.longDistance = longDistance;
+        this.x = x;
+        this.y = y;
+        this.z = z;
+        this.offsetX = offsetX;
+        this.offsetY = offsetY;
+        this.offsetZ = offsetZ;
+        this.maxSpeed = maxSpeed;
+        this.particleCount = particleCount;
+        this.data = data;
+    }
+
+    private ParticlePacket(ParticlePacket copy) {
+        this(copy.particleId, copy.longDistance, copy.x, copy.y, copy.z, copy.offsetX, copy.offsetY, copy.offsetZ, copy.maxSpeed, copy.particleCount, copy.data);
+    }
+
     public ParticlePacket(@NotNull NetworkBuffer reader) {
-        this(reader.read(VAR_INT), reader.read(BOOLEAN),
-                reader.read(DOUBLE), reader.read(DOUBLE), reader.read(DOUBLE),
-                reader.read(FLOAT), reader.read(FLOAT), reader.read(FLOAT),
-                reader.read(FLOAT), reader.read(INT), reader.read(RAW_BYTES));
+        this(readPacket(reader));
+    }
+
+    public ParticlePacket(Particle particle, boolean longDistance, double x, double y, double z, int offsetX, int offsetY, int offsetZ, int maxSpeed, int particleCount) {
+        this(particle.id(), longDistance, x, y, z, offsetX, offsetY, offsetZ, maxSpeed, particleCount, particle.data());
+    }
+
+    public ParticlePacket(Particle particle, double x, double y, double z, int offsetX, int offsetY, int offsetZ, int maxSpeed, int particleCount) {
+        this(particle.id(), false, x, y, z, offsetX, offsetY, offsetZ, maxSpeed, particleCount, particle.data());
+    }
+
+    private static ParticlePacket readPacket(NetworkBuffer reader) {
+        int particleId = reader.read(VAR_INT);
+        Boolean longDistance = reader.read(BOOLEAN);
+        Double x = reader.read(DOUBLE);
+        Double y = reader.read(DOUBLE);
+        Double z = reader.read(DOUBLE);
+        Float offsetX = reader.read(FLOAT);
+        Float offsetY = reader.read(FLOAT);
+        Float offsetZ = reader.read(FLOAT);
+        Float maxSpeed = reader.read(FLOAT);
+        Integer particleCount = reader.read(INT);
+        ParticleData data = ParticleData.read(particleId, reader);
+
+        return new ParticlePacket(particleId, longDistance, x, y, z, offsetX, offsetY, offsetZ, maxSpeed, particleCount, data);
     }
 
     @Override
     public void write(@NotNull NetworkBuffer writer) {
+        if (data != null && !data.validate(particleId)) throw new IllegalStateException("Particle data is not valid for this particle type");
```

Comment:
use `Check.stateCondition` (or maybe arg its more or less an argument), and provide the data type and particle type in the error message

Reply:
Fixed. Going to to use stateCondition, since argument feels like something that should be thrown when creating something